README.md
# IPIP

[![Gem Version](https://badge.fury.io/rb/ipip.svg)](http://badge.fury.io/rb/ipip)
[![Build Status](https://travis-ci.org/liluo/ipip.svg?branch=master)](https://travis-ci.org/liluo/ipip)

IP search based on [IPIP.net](http://ipip.net/)

### Getting Started

Install the gem:

```
gem install ipip
```

Then use it in your application:

```
require 'ipip'
                                # => Format: "国家\t省份或直辖市\t地区或城市"

IPIP.find('127.0.0.1')          # => "本机地址\t本机地址\t\t"

IPIP.find('211.147.4.49')       # => "中国\t北京\t北京\t"

IPIP.find('202.195.161.30')     # => "中国\t江苏\t镇江\t"

IPIP.find('www.douban.com')     # => "中国\t北京\t北京\t"

```

If you have `DATX` file(付费用户):

```
require 'ipip'

IPIPX.load_data!("/path/ipip.datx")

# => Format: "国家\t省份或直辖市\t地区或城市\t学校或单位\t运营商\t纬度\t经度\t时区1\t时区2\t行政代码.."

IPIPX.find('118.28.8.8')        # => "中国\t天津\t天津\t\t鹏博士\t39.125596\t117.190182\tAsia/Shanghai\tUTC+8\t120000\t86\tCN\tAP"

IPIPX.find("58.82.112.15")      # => "中国\t广东\t广州\t\t电信\t23.125178\t113.280637\tAsia/Shanghai\tUTC+8\t440100\t86\tCN\tAP"

IPIPX.find("60.191.8.80")       # => "中国\t浙江\t杭州\t\t电信\t30.287459\t120.153576\tAsia/Shanghai\tUTC+8\t330100\t86\tCN\tAP"

IPIPX.find("114.114.114.114")   # => "114DNS\t114DNS\t\tgreatbit.com\t\t\t\t\t\t\t\t*\t*"

```

You can try running `ipip [ip | domain]`:

```
$ ipip 202.195.161.30
Location for 202.195.161.30: 中国       江苏    镇江

$ ipip douban.com
Location for douban.com: 中国   北京    北京
```

### Contributing

* Fork the repository.
* Create a topic branch.
* Implement your feature or bug fix.
* Add, commit, and push your changes.
* Submit a pull request.

### ChangeLog

[ChangeLog](CHANGES.md)

### License

Released under the [MIT license](LICENSE).


CHANGES.md
## IPIP Changelog

#### v0.1.3 [2017-03-14]

- Bugfix: cannot find ip > 220.*(IPIPX)

#### v0.1.2 [2017-02-23]

- Remove `require 'pry'`

#### v0.1.1 [2017-02-23]

- Bugfix: raise `undefined method "encode" for nil:NilClass` when record not found

#### v0.1.0 [2017-02-23]

- Support `DATX` (付费数据格式)
- Upgrade 17monipdb.dat(201701)

#### v0.0.3 [2015-09-27]

- Upgrade 17monipdb.dat(201606)

#### v0.0.2 [2015-10-20]

- Upgrade 17monipdb.dat(201510)

#### v0.0.1 [2015-09-27]

- Release first version.


spec/ipip_spec.rb
# coding: utf-8

RSpec.describe IPIP do

  describe '#find' do

    shared_examples 'should return location' do |ip, location|
      it "should return #{location} when find #{ip}" do
        expect(IPIP.find(ip)).to eql location
      end
    end

    it_behaves_like 'should return location', '127.0.0.1', "本机地址\t本机地址\t\t"

    it_behaves_like 'should return location', '8.8.8.8', "GOOGLE\tGOOGLE\t\t"
    it_behaves_like 'should return location', '114.114.114.114', "114DNS\t114DNS\t\t"

    it_behaves_like 'should return location', '211.147.4.49', "中国\t北京\t北京\t"
    it_behaves_like 'should return location', '202.195.161.30', "中国\t江苏\t镇江\t"
    it_behaves_like 'should return location', '192.30.252.129', "美国\t美国\t\t"

    it_behaves_like 'should return location', 'douban.com', "中国\t北京\t北京\t"

  end
end


spec/spec_helper.rb
require 'ipip'

# This file was generated by the `rspec --init` command. Conventionally, all
# specs live under a `spec` directory, which RSpec adds to the `$LOAD_PATH`.
# The generated `.rspec` file contains `--require spec_helper` which will cause
# this file to always be loaded, without a need to explicitly require it in any
# files.
#
# Given that it is always loaded, you are encouraged to keep this file as
# light-weight as possible. Requiring heavyweight dependencies from this file
# will add to the boot time of your test suite on EVERY test run, even for an
# individual file that may not need all of that loaded. Instead, consider making
# a separate helper file that requires the additional dependencies and performs
# the additional setup, and require it from the spec files that actually need
# it.
#
# The `.rspec` file also contains a few flags that are not defaults but that
# users commonly want.
#
# See http://rubydoc.info/gems/rspec-core/RSpec/Core/Configuration
RSpec.configure do |config|
  # rspec-expectations config goes here. You can use an alternate
  # assertion/expectation library such as wrong or the stdlib/minitest
  # assertions if you prefer.
  config.expect_with :rspec do |expectations|
    # This option will default to `true` in RSpec 4. It makes the `description`
    # and `failure_message` of custom matchers include text for helper methods
    # defined using `chain`, e.g.:
    #     be_bigger_than(2).and_smaller_than(4).description
    #     # => "be bigger than 2 and smaller than 4"
    # ...rather than:
    #     # => "be bigger than 2"
    expectations.include_chain_clauses_in_custom_matcher_descriptions = true
  end

  # rspec-mocks config goes here. You can use an alternate test double
  # library (such as bogus or mocha) by changing the `mock_with` option here.
  config.mock_with :rspec do |mocks|
    # Prevents you from mocking or stubbing a method that does not exist on
    # a real object. This is generally recommended, and will default to
    # `true` in RSpec 4.
    mocks.verify_partial_doubles = true
  end

# The settings below are suggested to provide a good initial experience
# with RSpec, but feel free to customize to your heart's content.
=begin
  # These two settings work together to allow you to limit a spec run
  # to individual examples or groups you care about by tagging them with
  # `:focus` metadata. When nothing is tagged with `:focus`, all examples
  # get run.
  config.filter_run :focus
  config.run_all_when_everything_filtered = true

  # Limits the available syntax to the non-monkey patched syntax that is
  # recommended. For more details, see:
  #   - http://myronmars.to/n/dev-blog/2012/06/rspecs-new-expectation-syntax
  #   - http://teaisaweso.me/blog/2013/05/27/rspecs-new-message-expectation-syntax/
  #   - http://myronmars.to/n/dev-blog/2014/05/notable-changes-in-rspec-3#new__config_option_to_disable_rspeccore_monkey_patching
  config.disable_monkey_patching!

  # This setting enables warnings. It's recommended, but in some cases may
  # be too noisy due to issues in dependencies.
  config.warnings = true

  # Many RSpec users commonly either run the entire suite or an individual
  # file, and it's useful to allow more verbose output when running an
  # individual spec file.
  if config.files_to_run.one?
    # Use the documentation formatter for detailed output,
    # unless a formatter has already been configured
    # (e.g. via a command-line flag).
    config.default_formatter = 'doc'
  end

  # Print the 10 slowest examples and example groups at the
  # end of the spec run, to help surface which specs are running
  # particularly slow.
  config.profile_examples = 10

  # Run specs in random order to surface order dependencies. If you find an
  # order dependency and want to debug it, you can fix the order by providing
  # the seed, which is printed after each run.
  #     --seed 1234
  config.order = :random

  # Seed global randomization in this process using the `--seed` CLI option.
  # Setting this allows you to use `--seed` to deterministically reproduce
  # test failures related to randomization by passing the same `--seed` value
  # as the one that triggered the failure.
  Kernel.srand config.seed
=end
end


spec/ipip/ipv4_database_spec.rb
# coding: utf-8

RSpec.describe IPIP::IPv4Database do

  describe '.find' do

    let(:ipv4_database) { IPIP::IPv4Database.new }

    context 'when ip v4' do
      shared_examples 'should return location' do |ip, location|
        it "should return #{location} when find #{ip}" do
          expect(ipv4_database.find(ip)).to eql location
        end
      end

      it_behaves_like 'should return location', '127.0.0.1', "本机地址\t本机地址\t\t"

      it_behaves_like 'should return location', '8.8.8.8', "GOOGLE\tGOOGLE\t\t"
      it_behaves_like 'should return location', '114.114.114.114', "114DNS\t114DNS\t\t"

      it_behaves_like 'should return location', '211.147.4.49', "中国\t北京\t北京\t"
      it_behaves_like 'should return location', '202.195.161.30', "中国\t江苏\t镇江\t"

      it_behaves_like 'should return location', '192.30.252.129', "美国\t美国\t\t"
      it_behaves_like 'should return location', '225.109.248.81', ""
    end

    context 'when domain' do
      shared_examples 'should return location' do |domain|
        it "should return location when find #{domain}" do
          expect(ipv4_database.find(domain)).not_to be_nil
        end
      end

      it_behaves_like 'should return location', 'google.com'
      it_behaves_like 'should return location', 'douban.com'
    end

    context 'when ip v6' do
      shared_examples 'should return nil' do |ip|
        it "should return nil when find #{ip}" do
          expect(ipv4_database.find(ip)).to be_nil
        end
      end

      it_behaves_like 'should return nil', '2404:6800:4005:800::200e'
      it_behaves_like 'should return nil', '2402:f000:1:404:166:111:4:100'
    end

  end

end


lib/ipip.rb
require 'ipip/version'
require 'ipip/ipv4_database'
require 'ipip/ipv4_database_x'


module IPIP

  class << self

    def find(ip)
      ipv4_database.find(ip)
    end

    def load_data!(data_file)
      raise 'No such file' unless File.exists?(data_file)

      ipv4_database(data_file)
    end

    def ipv4_database(data_file = nil)
      @ipv4_database ||= IPv4Database.new(data_file)
    end

  end

end


module IPIPX

  class << self

    def find(ip)
      raise 'pls call `IPIPX.load_data!` first' unless data_loaded?

      ipv4_database.find(ip)
    end

    def load_data!(data_file)
      raise 'No such file' unless File.exists?(data_file)

      ipv4_database(data_file)
      @data_was_loaded = true
    end

    def ipv4_database(data_file = nil)
      @ipv4_database ||= IPv4DatabaseX.new(data_file)
    end

    def data_loaded?
      @data_was_loaded ||= false
    end

  end

end


lib/ipip/ipv4_database.rb
require 'socket'
require 'ipaddr'

module IPIP

  # Database for search IPv4 address.
  #
  #  The IPIP.net data file(17monipdb.dat) format in bytes::
  #
  #     ---------------------------
  #     | 4 bytes                 |             <- offset number
  #     ---------------------------
  #     | 256 * 4 bytes           |             <- first ip number index
  #     ---------------------------
  #     | offset - 1024 - 4 bytes |             <- ip index
  #     ---------------------------
  #     |    data  storage        |
  #     ---------------------------
  #
  class IPv4Database

    OFFSET_NUMBER_SIZE = 4
    FIRST_IP_NUMBER_INDEX_SIZE = 256 * 4
    IP_BLOCK_SIZE = 8
    DEFAULT_DATA_FILE = File.expand_path("../17monipdb.dat", __FILE__)

    def initialize(data_file = nil)
      @data_file = data_file || DEFAULT_DATA_FILE
    end

    def find(ip)
      ip = IPSocket::getaddress(ip)
      lookup_ipv4(ip)
    end

    private

    def offset
      @offset ||= unpack_N(binread(OFFSET_NUMBER_SIZE, 0))
    end

    def buffer
      @buffer ||= binread(nil, OFFSET_NUMBER_SIZE)
    end

    def binread(length, offset)
      IO.binread(@data_file, length, offset)
    end

    def lookup_ipv4(ip)
      ip = IPAddr.new(ip)
      return unless ip.ipv4?

      nip = ip.hton

      first_ip = unpack_C(nip)
      first_ip_offset = first_ip * 4

      start = unpack_V(buffer[first_ip_offset, 4])
      start_offset = start * IP_BLOCK_SIZE + FIRST_IP_NUMBER_INDEX_SIZE

      lo, hi = 0, (offset - start_offset) / IP_BLOCK_SIZE

      while lo < hi
        mid = (lo + hi) / 2
        mid_offset = start_offset + mid * IP_BLOCK_SIZE
        mid_val = buffer[mid_offset, 4]

        if mid_val < nip
          lo = mid + 1
        else
          hi = mid
        end
      end

      start_offset += lo * IP_BLOCK_SIZE
      return if start_offset == offset

      data_pos = unpack_V(buffer[start_offset + 4, 3] + 0.chr)
      data_length = unpack_C(buffer[start_offset + 7])

      data_offset = offset + data_pos - FIRST_IP_NUMBER_INDEX_SIZE - OFFSET_NUMBER_SIZE
      data = buffer[data_offset, data_length].to_s.encode('UTF-8', 'UTF-8')
    end

    [:C, :N, :V].each do |format|
      define_method "unpack_#{format}" do |binary|
        binary.unpack(format.to_s).first
      end
    end
  end
end


lib/ipip/ipv4_database_x.rb
require 'socket'
require 'ipaddr'

module IPIPX

  # Database for search IPv4 address.
  #
  #  The IPIP.net data file(ipip.datx) format in bytes::
  #
  #     ----------------------------------
  #     | 4 bytes                        |             <- offset number
  #     ----------------------------------
  #     | 256 * 1024 bytes               |             <- first ip number index
  #     ----------------------------------
  #     | offset - 256 * 1024 - 4 bytes  |             <- ip index
  #     ----------------------------------
  #     | data  storage                  |
  #     ----------------------------------
  #
  class IPv4DatabaseX

    OFFSET_NUMBER_SIZE = 4
    FIRST_IP_NUMBER_INDEX_SIZE = 256 * 1024
    IP_BLOCK_SIZE = 9

    def initialize(data_file)
      @data_file = data_file
    end

    def find(ip)
      ip = IPSocket::getaddress(ip)
      lookup_ipv4(ip)
    end

    private

    def offset
      @offset ||= unpack_N(binread(OFFSET_NUMBER_SIZE, 0))
    end

    def buffer
      @buffer ||= binread(nil, OFFSET_NUMBER_SIZE)
    end

    def binread(length, offset)
      IO.binread(@data_file, length, offset)
    end

    def lookup_ipv4(ip)
      ip = IPAddr.new(ip)
      return unless ip.ipv4?

      nip = ip.hton
      first_ip = unpack_C(nip)
      ipdot = ip.to_s.split('.')
      first_ip_offset = (first_ip * 256 + ipdot[1].to_i) * 4

      start = unpack_V(buffer[first_ip_offset, 4])
      start_offset = start * IP_BLOCK_SIZE + FIRST_IP_NUMBER_INDEX_SIZE
      end_offset = offset - FIRST_IP_NUMBER_INDEX_SIZE - OFFSET_NUMBER_SIZE
      lo, hi = 0, (end_offset - start_offset) / IP_BLOCK_SIZE

      while lo < hi
        mid = (lo + hi) / 2
        mid_offset = start_offset + mid * IP_BLOCK_SIZE
        mid_val = buffer[mid_offset, 4]

        if mid_val < nip
          lo = mid + 1
        else
          hi = mid
        end
      end

      start_offset += lo * IP_BLOCK_SIZE
      return if start_offset == offset

      data_pos = unpack_V(buffer[start_offset + 4, 3] + 0.chr)
      data_length = unpack_C(buffer[start_offset + 8])

      data_offset = offset + data_pos - FIRST_IP_NUMBER_INDEX_SIZE - OFFSET_NUMBER_SIZE
      data = buffer[data_offset, data_length].to_s.encode('UTF-8', 'UTF-8')
    end

    [:C, :N, :V].each do |format|
      define_method "unpack_#{format}" do |binary|
        binary.unpack(format.to_s).first
      end
    end
  end
end
